_posts/2020-06-28-openfaas-loki-into.md

```diff
@@ -0,0 +1,235 @@
+
+---
+
+In this post Lucas will introduce the [`openfaas-loki`](https://github.com/LucasRoesler/openfaas-loki) [log provider](https://docs.openfaas.com/architecture/logs-provider/) and how to use it to get historic logs for your OpenFaaS functions.
+
+In any production service, logs are a critical component of observability and debuggability of the system. OpenFaaS builds in support for logs in _every_ installation. But it can be even better by using a new log provider. In this post we will show you how to supercharge the logs by installing [Grafana's Loki](https://grafana.com/oss/loki) and configure OpenFaaS to use Loki.
+
+## Motivation
+
+In the default installation of OpenFaaS, the logs come from the underlying orchestration tool. This is great, because logs are automatically available without any additional effort, but this means logs only exist for the _current_ function instances. This means that as your function scales from 0 to 100 and then back to 0, you will no long have access to the function logs because the orchestraition system does not keep logs after the function instance is removed! This is simply the default behavior of systems like Docker Swarm and Kubernetes.  If a user reports and error, but the function instance has scaled down already, the logs you need to debug the issue are now gone :(
+
+To have logs for long ago terminated function instances, we need to install some kind of logs aggregator. Traditionally, this has meant the [ELK stack](https://www.elastic.co/what-is/elk-stack): ElasticSearch, Logstash, Kibana. This is a great and well established tool; but, in [late 2018 Grafana introduced](https://kccna18.sched.com/event/GrXC/on-the-oss-path-to-full-observability-with-grafana-david-kaltschmidt-grafana-labs) a new alternative aimed squarely at the Cloud Native world: [Loki](https://grafana.com/oss/loki/).
+
+> Loki is a horizontally-scalable, highly-available, multi-tenant log aggregation system inspired by Prometheus. It is designed to be very cost effective and easy to operate. It does not index the contents of the logs, but rather a set of labels for each log stream.
+
+Loki is a light weight, easy to install, very easy to use. It is especially easy to use in Kuberenetes.
+
+Once a log aggregator is installed and configured it will often have a UI for exploring the logs, for ELK this is [Kibana](https://www.elastic.co/kibana) and for Loki it is [Grafana Explore](https://grafana.com/docs/grafana/latest/features/explore/). However, since OpenFaaS Gateway 0.15.0+ (July 2019), OpenFaaS supports [pluggable logging providers](https://docs.openfaas.com/architecture/logs-provider/) to support streaming function logs via the Gateway API and the `faas-cli`.
+
+The simplicy of Loki made it easy to turn into the first external [log provider](https://docs.openfaas.com/architecture/logs-provider/) for OpenFaaS that can now be easily installed using [arkade](https://get-arkade.dev/), which we will walk-though below.
```

Comment:
> The simplicy of Loki made it easy to turn into the first external [log provider](https://docs.openfaas.com/architecture/logs-provider/) for OpenFaaS that can now be easily installed using [arkade](https://get-arkade.dev/), which we will walk-though below.

Rather than talking in the past participle, could we relate this to your experience (you as the author)?

When I wrote the OpenFaaS logs provider for Loki, I found the API easy to work with and the simplicity of the project meant the whole thing was completed in a short period of time.

Reply:
I will just remove the sentence. I don't like how the active versions sound and it isn't important enough to try and fight for it